torch/testing/_internal/common_utils.py

```diff
@@ -1767,304 +1912,77 @@ def compare_with_numpy(self, torch_fn, np_fn, tensor_like,
-            else:
-                self.assertEqual(set(x.keys()), set(y.keys()), atol=atol, rtol=rtol,
-                                 msg=msg, exact_dtype=exact_dtype,
-                                 exact_device=exact_device)
-                key_list = list(x.keys())
-                self.assertEqual([x[k] for k in key_list],
-                                 [y[k] for k in key_list],
-                                 atol=atol, rtol=rtol, msg=msg,
-                                 exact_dtype=exact_dtype, exact_device=exact_device)
-        elif isinstance(x, type) and isinstance(y, type):
-            # See TestTorch.test_assert_equal_generic_meta
-            debug_msg = ("Attempted to compare [type] types: "
-                         f"Expected: {x}; Actual: {y}.")
-            super().assertEqual(x, y, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
-        elif is_iterable(x) and is_iterable(y):
-            debug_msg = ("Attempted to compare the lengths of [iterable] types: "
-                         f"Expected: {len(x)}; Actual: {len(y)}.")
-            super().assertEqual(len(x), len(y), msg=self._get_assert_msg(msg, debug_msg=debug_msg))
-            for x_, y_ in zip(x, y):
-                self.assertEqual(x_, y_, atol=atol, rtol=rtol, msg=msg,
-                                 exact_dtype=exact_dtype, exact_device=exact_device)
-        elif isinstance(x, bool) and isinstance(y, bool):
-            super().assertTrue(x == y, msg=msg)
-
-        # Scalar x Scalar
-        elif isinstance(x, Number) and isinstance(y, Number):
-            result, debug_msg_scalars = self._compareScalars(x, y, rtol=rtol, atol=atol,
-                                                             equal_nan=equal_nan)
-            if not result:
-                assert debug_msg_scalars is not None
-                debug_msg = "Scalars failed to compare as equal! " + debug_msg_scalars
-            super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
-        else:
-            super().assertEqual(x, y, msg=msg)
+    def assertEqual(
+            self,
+            x,
+            y,
+            msg: Optional[str] = None,
+            *,
+            atol: Optional[float] = None,
+            rtol: Optional[float] = None,
+            equal_nan=True,
+            exact_dtype=True,
+            exact_device=False,
+            exact_layout=False,
+            exact_stride=False,
+            exact_is_coalesced=False
+    ):
+        # Hide this function from `pytest`'s traceback
+        __tracebackhide__ = True
+
+        if any(
+            isinstance(input, np.ndarray) and not has_corresponding_torch_dtype(input.dtype) for input in (x, y)
+        ):
+            def to_list(input):
+                return input.tolist() if isinstance(input, (torch.Tensor, np.ndarray)) else list(input)
+
+            x = to_list(x)
+            y = to_list(y)
+        elif isinstance(x, torch.Tensor) and isinstance(y, Sequence):
```

Comment:
Comments here too

Readability nit: maybe keep `x` on the left and `y` on the right? 

Follow-up idea: `actual` and `expected` might be more readable names than `x` and `y` but nbd

Reply:
Agreed. As long as no one is doing `self.assertEqual(x=..., y=...)` we can rename. I'll try in a follow-up PR.